### PR TITLE
Fact store lookup

### LINKED ISSUE #35: Enable custom data for (re)training
Add ability to provide user defined data.

E.g user walks through entries marked as anomalies and provides false positives back. We need to be able   to add these manually reviewed entries into training data to make sure false positives get lower score in the future.

### PR DESCRIPTION
**Pull request by zmhassan**
### Description
We have this feature done. We are now able to run a training run and inject labelled false anomaly data from fact_store. I'll be deploying this to PR once you'll review and merge. Friday is deadline but why not get it in earlier. @durandom 

This PR includes code to:
1. Inject false positive data to training run 
2. Filter false positive messages that appear so we avoid users from getting notified of reported false anomalies.

## Steps to run locally:
Note: Make sure you set SQL_CONNECT env var to mysql. Preload it with data if you want. Then start fact_store:
`python app.py run ui`
Run training and inference 
`python app.py run --config-yaml .test_env_config_fact_store.yaml`

Fixes 
#35

### CODE REVIEW

File: .test_env_config_fact_store.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+STORAGE_BACKEND: "local"
+LS_INPUT_PATH: "validation_data/log_anomaly_detector-100000-events.json"
+LS_OUTPUT_PATH: "validation_data/local-results-1.txt"
+INFER_ANOMALY_THRESHOLD: 1.3
+INFER_TIME_SPAN: 1
+INFER_LOOPS: 1
+PARALLELISM: 6
+FREQ_NOISE: 10000
```
Comment (durandom) [inline .test_env_config_fact_store.yaml]: can we add the new settings, like `FREQ_NOISE` to the readme to explain what they are doing?

File: .test_env_config_fact_store.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+STORAGE_BACKEND: "local"
```
Comment (durandom) [inline .test_env_config_fact_store.yaml]: How is this file `.test_env_config_fact_store.yaml` used? Can't we re-use the exisiting `.test_env_config.yaml`?

**Comment by durandom** [comment]
Oh, I see some answers now in the description.

How about adding a section to the readme on how to run it? Otherwise that info is lost in this PR

File: .test_env_config_fact_store.yaml (hunk 2 above)
Comment (zmhassan) [reply .test_env_config_fact_store.yaml]: I think we should keep this additional file. Its an example of config that someone would use if say they wanted to run fact store locally and have anomaly detector work with it. I think we should keep it.

File: .test_env_config_fact_store.yaml (hunk 1 above)
Comment (zmhassan) [reply .test_env_config_fact_store.yaml]: The purpose is that we want to increase the occurrences of log lines that are false positives so the SOM model will give it a lower score thus not triggering an false anomaly prediction. We have an additional check in place. I guess we can probably measure how well this performs overtime. W2v and SOM ML Models should learn overtime  Anomaly_False_Positive

**Comment by zmhassan** [comment]
I'll add more details to the README.md

File: .test_env_config_fact_store.yaml (hunk 1 above)
Comment (durandom) [reply .test_env_config_fact_store.yaml]: good. add that explanation to the readme, where the other variables are explained

File: .test_env_config_fact_store.yaml (hunk 2 above)
Comment (durandom) [reply .test_env_config_fact_store.yaml]: ok. same, add the way to execute this to the README.md 
or, even better, add a `Makefile` that would execute this and other ways to run tests etc.
Just like here: https://github.com/AICoE/prometheus-flatliner/blob/master/Makefile#L61

File: anomaly_detector/anomaly_detector.py [hunk 3]
```diff
@@ -215,11 +222,28 @@ def infer(self):
         # When we reached # of inference loops, retrain models
         self.recreate_models = False
 
+    def fetch_false_positives(self):
+        """Fetch false positive from datastore and add noise to training run."""
+        _LOGGER.info("Fetching false positives from fact store")
+        try:
+            r = requests.get(url=self.config.FACT_STORE_URL + "/api/false_positive")
+            data = r.json()
+            data["feedback"]
```
Comment (vpavlin) [inline anomaly_detector/anomaly_detector.py]: What is this for?

File: anomaly_detector/anomaly_detector.py [hunk 4]
```diff
@@ -215,11 +222,28 @@ def infer(self):
...
+            fp = []
+            for msg in data["feedback"]:
+                noise = [{"message": msg}] * self.config.FREQ_NOISE
+                fp.extend(noise)
+            _LOGGER.info("Added noise {} messages ".format(len(fp)))
+            return fp
+        except Exception as ex:
+            _LOGGER.debug("Fact Store is either down or not functioning")
```
Comment (vpavlin) [inline anomaly_detector/anomaly_detector.py]: Should the used log level for the message ber `error`? There is an expection after all and you are reporting it as `debug` here which would not ever get printed in prod environment

File: anomaly_detector/anomaly_detector.py [hunk 5]
```diff
@@ -215,11 +222,28 @@ def infer(self):
...
+            return None
+
     def run(self, single_run=False):
         """Run the main loop."""
         start_http_server(8080)
         srun = False
-        fp = None
+        fp = self.fetch_false_positives()
         while True and srun is False:
```
Comment (vpavlin) [inline anomaly_detector/anomaly_detector.py]: `True` is unnecessary here as it's always true:) and the second part of the condition decides anyway

File: anomaly_detector/fact_store/fact_store_api.py [hunk 6]
```diff
@@ -33,6 +34,7 @@ def write_event(self, predict_id, message, score, anomaly_status):
         event = EventModel(message=message, score=score, predict_id=predict_id, anomaly_status=anomaly_status)
         self.session.add(event)
         self.session.commit()
+        return True
         print("Event ID: {}  recorded in events_store".format(event.predict_id))
```
Comment (vpavlin) [inline anomaly_detector/fact_store/fact_store_api.py]: The code will never get to the `print`, you can remove it or move above return ...also probably should be `_LOGGER.debug`

File: anomaly_detector/fact_store/fact_store_api.py [hunk 7]
```diff
@@ -57,3 +59,13 @@ def readall_feedback(self):
         feedbacks = self.session.query(FeedbackModel).all()
         list = [f.to_dict() for f in feedbacks]
         return list
+
+    def readall_false_positive(self):
+        """Service for querying datastore of current false anomalies."""
+        items = self.session.query(FeedbackModel).all()
+        messages = set()
+        for i in items:
+            events = self.session.query(EventModel).filter_by(predict_id=i.predict_id)
```
Comment (vpavlin) [inline anomaly_detector/fact_store/fact_store_api.py]: I did not review current version of the model, but I assume `predict_id` is primary key for EventModel and foreign key in FeedbackModel..you could probably use an SQL join and skip the for loop

File: anomaly_detector/anomaly_detector.py [hunk 8]
```diff
@@ -78,11 +79,12 @@ def _load_data(self, time_span, max_entries, fp=None):
         return data, raw
 
     @TRAINING_TIME.time()
-    # @profile
     def train(self, fp=None, node_map=24):
         """Train models for anomaly detection."""
         start = time.time()
         data, _ = self._load_data(self.config.TRAIN_TIME_SPAN, self.config.TRAIN_MAX_ENTRIES, fp)
+        if fp is not None:
```
Comment (vpavlin) [inline anomaly_detector/anomaly_detector.py]: I would also suggest renaming `fp` to `false_positives` as `fp` is usually used as "file pointer" and might be confusing for new contributors

File: openshift/factstore.app.yaml [hunk 9]
```diff
@@ -0,0 +1,87 @@
+apiVersion: template.openshift.io/v1
```
Comment (vpavlin) [inline openshift/factstore.app.yaml]: This file should not be part of this PR as there is a separate PR for it. Or you should close that PR in favour of this one

File: anomaly_detector/anomaly_detector.py [hunk 10]
```diff
@@ -215,11 +222,28 @@ def infer(self):
         # When we reached # of inference loops, retrain models
         self.recreate_models = False
 
+    def fetch_false_positives(self):
+        """Fetch false positive from datastore and add noise to training run."""
+        _LOGGER.info("Fetching false positives from fact store")
+        try:
+            r = requests.get(url=self.config.FACT_STORE_URL + "/api/false_positive")
```
Comment (vpavlin) [inline anomaly_detector/anomaly_detector.py]: It would be good to check StatusCode from the response

File: openshift/factstore.app.yaml (hunk 9 above)
Comment (zmhassan) [reply openshift/factstore.app.yaml]: I've removed this from this PR.

File: anomaly_detector/fact_store/fact_store_api.py (hunk 6 above)
Comment (zmhassan) [reply anomaly_detector/fact_store/fact_store_api.py]: Good catch I'll move this down

File: anomaly_detector/anomaly_detector.py (hunk 3 above)
Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: It shouldn't be there. Typo.

Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: Removed.

File: anomaly_detector/anomaly_detector.py (hunk 4 above)
Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: Woops. Thanks... Good catch.

Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: Fixed

File: anomaly_detector/anomaly_detector.py (hunk 5 above)
Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: Thats correct. Thank you. Fixed it.

File: anomaly_detector/fact_store/fact_store_api.py (hunk 6 above)
Comment (zmhassan) [reply anomaly_detector/fact_store/fact_store_api.py]: FIxed.

File: anomaly_detector/fact_store/fact_store_api.py (hunk 7 above)
Comment (zmhassan) [reply anomaly_detector/fact_store/fact_store_api.py]: > I did not review current version of the model, but I assume `predict_id` is primary key for EventModel and foreign key in FeedbackModel..you could probably use an SQL join and skip the queries in th for loop

I'm planning on doing some changes to the schema. For now this works. I've opened a git issue based on your comment.

File: anomaly_detector/anomaly_detector.py (hunk 8 above)
Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: Done

File: anomaly_detector/anomaly_detector.py (hunk 10 above)
Comment (zmhassan) [reply anomaly_detector/anomaly_detector.py]: We will be moving this to GRPC so the rest http stuff is going to be removed in future PR. I will open a git issue for now.

File: .test_env_config_fact_store.yaml (hunk 2 above)
Comment (zmhassan) [reply .test_env_config_fact_store.yaml]: Yup I added a make file to this PR.

File: .test_env_config_fact_store.yaml (hunk 1 above)
Comment (zmhassan) [reply .test_env_config_fact_store.yaml]: Done

**Comment by zmhassan** [comment]
[Retraining of LAD based on user feedback [OKR5][KR5]](https://trello.com/c/EESYvI4p/247-retraining-of-lad-based-on-user-feedback-okr5kr5)
